Post for tidymodels Q2 update (#518)
tidyverse/tidyverse.org: content/blog/tidymodels-2021-q2/index.Rmd

---
output: hugodown::hugo_document

slug: tidymodels-2021-q2

> title: Q2 tidymodels digest

Reviewer: Include year in title?

Author: Great point 👍

- [stacks](https://github.com/tidymodels/stacks/blob/main/NEWS.md#v020)
- [themis](https://themis.tidymodels.org/dev/news/index.html#themis-0-1-4-2021-06-12)
- [tidymodels](https://tidymodels.tidymodels.org/news/index.html#tidymodels-0-1-3-2021-04-19)
- [tidyposterior](https://github.com/tidymodels/tidyposterior/blob/master/NEWS.md#tidyposterior-010)
- [tune](https://tune.tidymodels.org/news/index.html#tune-0-1-5-2021-04-23)
- [workflowsets](https://workflowsets.tidymodels.org/news/index.html#workflowsets-0-0-2-2021-04-16)
- [yardstick](https://yardstick.tidymodels.org/news/index.html#yardstick-0-0-8-2021-03-28)

## Reduce the memory footprint of your recipes

The [butcher](https://butcher.tidymodels.org/) package provides methods to remove (or "axe") components from model objects that are not needed for prediction. The most recent release updated how butcher handles _recipes_ (the tidymodels approach for preprocessing and feature engineering) for more complete and robust coverage. Let's consider a simulated churn-classification dataset for phone company customers:

```{r, fig.width=8, fig.height=5, message=FALSE}

> library(tidymodels)

Reviewer: I wonder why these aren't getting autolinked in the output

---
output: hugodown::hugo_document

slug: tidymodels-2021-q2
title: Q2 tidymodels digest
date: 2021-07-02
author: Julia Silge
description: >
    Releases of tidymodels packages in Q2 of 2021 include more streamlined

>     memory footprints for feature engineering recipes, new model engine options,

Reviewer 2: ```suggestion
    memory footprints for feature-engineering recipes, new model engine options,
```

if we are following @batpigandme's suggestions

    and better support for post-processing predictions.

photo:
  url: https://unsplash.com/photos/UsZNiW6LyQ0
  author: SHOT

# one of: "deep-dive", "learn", "package", "programming", or "other"
categories: [roundup] 
tags: [tidymodels, parsnip, recipes, butcher, yardstick, probably]
---

```{r include=FALSE}
library(ggplot2)
theme_set(theme_minimal())
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

<!--
TODO:
* [ ] Look over / edit the post's title in the yaml
* [ ] Edit (or delete) the description; note this appears in the Twitter card
* [ ] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
* [ ] Find photo & update yaml metadata
* [ ] Create `thumbnail-sq.jpg`; height and width should be equal
* [ ] Create `thumbnail-wd.jpg`; width should be >5x height
* [ ] `hugodown::use_tidy_thumbnails()`
* [ ] Add intro sentence, e.g. the standard tagline for the package
* [ ] `usethis::use_tidy_thanks()`
-->

> The [tidymodels](https://www.tidymodels.org/) framework is a collection of R packages for modeling and machine learning using tidyverse principles. Earlier this year, we [started regular updates](https://www.tidyverse.org/blog/2021/03/tidymodels-2021-q1/) here on the tidyverse blog summarizing recent developments in the tidymodels ecosystem overall. You can check out the [`tidymodels`](https://www.tidyverse.org/tags/tidymodels/) tag to find all tidymodels blog posts here, including those that focus on a single package or more major releases, but the purpose of these roundup posts is to keep you informed about any releases you may have missed and useful new functionality as we maintain these packages.

Reviewer 3: ```suggestion
The [tidymodels](https://www.tidymodels.org/) framework is a collection of R packages for modeling and machine learning using tidyverse principles. Earlier this year, we [started regular updates](https://www.tidyverse.org/blog/2021/03/tidymodels-2021-q1/) here on the tidyverse blog summarizing recent developments in the tidymodels ecosystem. You can check out the [`tidymodels tag`](https://www.tidyverse.org/tags/tidymodels/) to find all tidymodels blog posts here, including those that focus on a single package or more major releases. The purpose of these roundup posts is to keep you informed about any releases you may have missed and useful new functionality as we maintain these packages.
```

Reviewer 3: "overall" seemed redundant (if you want to keep it, maybe put it before "tidymodels ecosystem"?

I changed the linked text to be more descriptive of the link for accessibility (I am terrible at this, and often make the mistake of linking to the word *here* or something to that effect).

> Since our last roundup post, there have been 19 CRAN releases of 15 different packages. That might sound like a lot of change to absorb as a tidymodels user! However, we purposefully write code in small, modular packages that we can release frequently, to make models easier to deploy and our software easier to maintain. You can install these updates from CRAN with:

Reviewer 3: ```suggestion
Since our last roundup post, there have been 19 CRAN releases of 15 different packages. That might sound like a lot of change to absorb as a tidymodels user! However, we purposefully write code in small, modular packages that we can release frequently to make models easier to deploy and our software easier to maintain. You can install these updates from CRAN with:
```

Reviewer 3: Extraneous comma (aside: I add these little notes when I'm not sure how obvious the change will be in the GitHub UI—let me know if they're simply annoying, and I'll stop)

- [butcher](https://butcher.tidymodels.org/news/index.html#butcher-0-1-4-2021-03-19)
- [discrim](https://discrim.tidymodels.org/news/index.html#discrim-0-1-2-2021-05-28)
- [hardhat](https://hardhat.tidymodels.org/news/index.html#hardhat-0-1-4-2020-07-02)
- [parsnip](https://parsnip.tidymodels.org/news/index.html#parsnip-0-1-6-2021-05-27)
- [probably](https://probably.tidymodels.org/news/index.html#probably-0-0-6-2020-06-05)
- [recipes](https://recipes.tidymodels.org/news/index.html#recipes-0-1-16-2021-04-16)
- [rsample](https://rsample.tidymodels.org/news/index.html#rsample-0-1-0-2021-05-08)
- [stacks](https://github.com/tidymodels/stacks/blob/main/NEWS.md#v020)
- [themis](https://themis.tidymodels.org/dev/news/index.html#themis-0-1-4-2021-06-12)
- [tidymodels](https://tidymodels.tidymodels.org/news/index.html#tidymodels-0-1-3-2021-04-19)
- [tidyposterior](https://github.com/tidymodels/tidyposterior/blob/master/NEWS.md#tidyposterior-010)
- [tune](https://tune.tidymodels.org/news/index.html#tune-0-1-5-2021-04-23)
- [workflowsets](https://workflowsets.tidymodels.org/news/index.html#workflowsets-0-0-2-2021-04-16)

> - [yardstick](https://yardstick.tidymodels.org/news/index.html#yardstick-0-0-8-2021-03-28)

Reviewer 3: I don't have a hot take on this, but do you think it's worth adding the version number for each item here, since you're linking to the anchored item?

Author: I don't think that will add much for the reader? I think I'll pass on that.

## Reduce the memory footprint of your recipes

> The [butcher](https://butcher.tidymodels.org/) package provides methods to remove (or "axe") components from model objects that are not needed for prediction. There are many methods for various kinds of models, but the most recent release updated how butcher handled _recipes_, the tidymodels' approach for preprocessing and feature engineering, for more complete and robust coverage. Let's consider a simulated churn classification data set for phone company customers:

Reviewer 4: ```suggestion
The [butcher](https://butcher.tidymodels.org/) package provides methods to remove (or "axe") components from model objects that are not needed for prediction. There are many methods for various kinds of models, but the most recent release updated how butcher handles _recipes_, the tidymodels' approach for preprocessing and feature engineering, for more complete and robust coverage. Let's consider a simulated churn classification data set for phone company customers:
```

Reviewer 3: ```suggestion
The [butcher](https://butcher.tidymodels.org/) package provides methods to remove (or "axe") components from model objects that are not needed for prediction. There are many methods for various kinds of models, but the most recent release updated how butcher handled _recipes_ (the tidymodels approach for preprocessing and feature engineering) for more complete and robust coverage. Let's consider a simulated churn-classification dataset for phone company customers:
```

Reviewer 3: I think when you have the "the" in there for "the tidymodels approach" it's not possessive (it would be if it was "tidymodels' approach").

Also, in the following:
> There are many methods for various kinds of models

are you referring to literal R methods? I can't decide if it's one of those special words that we basically can't use in its more quotidian sense because of the context of "being about R." One option would be just to cut that first clause and begin with:
> The most recent release…

I don't think you'd lose much info.

"churn-classification" gets a hyphen just because it's modifying a noun (compound modifier)

Lastly, I think dataset has reached its evolution into closed-form (especially in the [technical context](https://english.stackexchange.com/a/2129/44117)). The two word (open form) -> compound word (hyphenated) -> one word (closed form) evolution is always up for debate (because, y'know, English be like that), but I go closed when it resolves ambiguity _especially_ when the latter term can also be parsed as a verb.

```{r, fig.width=8, fig.height=5, message=FALSE}
library(tidymodels)
library(butcher)
data("mlc_churn")

set.seed(123)
churn_split <- initial_split(mlc_churn)
churn_train <- training(churn_split)
churn_test  <- testing(churn_split)

ggplot(churn_train, aes(y = voice_mail_plan, fill = churn)) +
  geom_bar(alpha = 0.8, position = "fill") +
  scale_x_continuous(labels = scales::percent_format()) +
  labs(x = NULL)
```

> For some kinds of models, we would want to create dummy or indicator variables from nominal predictors and preprocess features to be on the same scale. We can use recipes for this task:

Reviewer 3: ```suggestion
For some kinds of models, we would want to create dummy or indicator variables from nominal predictors, and preprocess features to be on the same scale. We can use recipes for this task:
```

```{r}
churn_rec <- 
  recipe(churn ~ voice_mail_plan + total_intl_minutes + 
           total_day_minutes + total_eve_minutes + state, 
         data = churn_train) %>%
  step_dummy(all_nominal_predictors()) %>%
  step_normalize(all_predictors())
```

You can `prep(churn_rec)` to estimate the quantities needed to create categorical features and to scale all the predictors:

```{r}
churn_prep <- prep(churn_rec)
churn_prep
```

> To remove everything from this prepped recipe not needed for applying to new data (e.g. [bake()](https://recipes.tidymodels.org/reference/bake.html) it), we can call `butcher(churn_prep)`. In some applications, modeling practitioners need to make custom functions with a feature engineering recipe. Sometimes those functions have... a lot of extra STUFF in them, stuff that is not needed for prediction.

Reviewer 3: ```suggestion
To remove everything from this prepped recipe not needed for applying to new data (e.g. [bake()](https://recipes.tidymodels.org/reference/bake.html) it), we can call `butcher(churn_prep)`. In some applications, modeling practitioners need to make custom functions with a feature-engineering recipe. Sometimes those functions have... a lot of extra STUFF in them, stuff that is not needed for prediction.
```

Reviewer 3: compound modifier 🤓

```{r}
butchered_plus <- function() {
  some_junk_in_the_environment <- runif(1e6)
  
  churn_prep <- 
    recipe(churn ~ voice_mail_plan + total_intl_minutes + 
             total_day_minutes + total_eve_minutes + state, 
           data = churn_train) %>%
    step_dummy(all_nominal_predictors()) %>%
    step_normalize(all_predictors()) %>%
    prep()
  
  butcher(churn_prep)
}
```

In the old version of butcher, we did not successfully remove all that extra stuff, and recipes were bigger than they needed to be:

```{r, eval=FALSE}

> lobstr::obj_size(butcher(churn_prep))

Reviewer 3: I might add a comment above the code here giving the version number of butcher used, or at least say `# old version of butcher` and `# new version of butcher` (or something to that effect). My theory is that some readers pretty much look at the code exclusively, and might not catch the difference (esp since it's shown via `eval = FALSE`/not something they'll reproduce by running the code themselves).

#> 1,835,512 B
lobstr::obj_size(butchered_plus())
#> 9,836,480 B
```

In the new version of butcher, we now successfully remove unneeded components from the recipe, so it is smaller:

```{r, eval=FALSE}
lobstr::obj_size(butcher(churn_prep))
#> 1,695,352 B
lobstr::obj_size(butchered_plus())
#> 1,695,352 B
```

There are also `butcher()` methods for `workflow()` objects, so when you `butcher()` a modeling workflow, you remove everything not needed for prediction from both its estimated recipe *and* its trained model, making it as lightweight as possible for deployment.

## SVMs and fast logistic regression with LiblineaR

> Unfortunately, the `"liquidSVM"` engine for support vector machine models that parsnip supported was deprecated in the latest release, because that package was removed from CRAN. We added a new engine in the same release that allows users to fit linear SVMs with the [parsnip model `svm_linear()`](https://parsnip.tidymodels.org/reference/svm_linear.html), as well as having another option for logistic regression. This new `"LiblineaR"` engine is based on the same C++ library that is shipped with scikit-learn. We'd like to thank the [maintainers of the LiblineaR R package](https://www.dnalytics.com/software/liblinear/) for all their help as we set up this integration.

Reviewer 3: Add link to scikit-learn or the relevant C++ lib? (Totally your call—I think it's fine either way).

```{r}
set.seed(234)
churn_folds <- vfold_cv(churn_train, v = 5, strata = churn)

> liblinear_wf <-

Reviewer 4: ```suggestion
liblinear_spec <-
```

  logistic_reg(penalty = 0.2, mixture = 1) %>%
  set_mode("classification") %>%
  set_engine("LiblineaR")

liblinear_wf <-
  workflow() %>%
  add_recipe(churn_rec) %>%

>   add_model(liblinear_wf)

Reviewer 4: ```suggestion
  add_model(liblinear_spec)
```

fit_resamples(liblinear_wf, resamples = churn_folds)
```

The `"LiblineaR"` engine for regularized logistic regression [can be very fast compared to the `"glmnet"` engine](https://twitter.com/juliasilge/status/1374029310764359681), even when we use a [sparse representation](https://www.tidyverse.org/blog/2020/11/tidymodels-sparse-support/). Check out [benchmarking code here](https://gist.github.com/juliasilge/26a43e5e68cf12842354e6652dfed688).

> ## Post-processing your model predictions with probably and yardstick

Reviewer 3: A musing: It's so weird to me that preprocessing has definitively reached closed form, but post-processing is still in compound.

Author: YES, so weird!

We recently had releases of both the [yardstick](https://yardstick.tidymodels.org/) and [probably](https://probably.tidymodels.org/) packages, which now work together even better. The probably package can, among other things, help you post-process your model predictions. This data on churn is imbalanced, with many more customers who did not churn than those who did; we may need to use a threshold other than 0.5 for most appropriate results, or an organization may want to set a specific threshold for some action to prevent churn. You can set a threshold using the [probably function `make_two_class_pred()`](https://probably.tidymodels.org/reference/make_class_pred.html).

```{r}
library(probably)

set.seed(123)
churn_preds <- 
  liblinear_wf %>%
  fit(churn_train) %>%
  augment(churn_test)

churn_post <- 
  churn_preds %>%
  mutate(.pred = make_two_class_pred(.pred_yes, levels(churn), threshold = 0.7))
```

> The class predictions created with probably integrate well with functions from yardstick, including custom sets of metrics created with `metric_set()`.

Reviewer 3: ```suggestion
The class predictions created with probably integrate well with functions from yardstick, including custom sets of metrics created with [`metric_set()`](https://yardstick.tidymodels.org/reference/metric_set.html).
```
